pkg/component/pushed_component.go

```diff
@@ -0,0 +1,276 @@
+
+func (s s2iComponent) GetLinkedSecretNames() (secretNames []string) {
+	for _, env := range s.dc.Spec.Template.Spec.Containers[0].EnvFrom {
+		if env.SecretRef != nil {
+			secretNames = append(secretNames, env.SecretRef.Name)
+		}
+	}
+	return secretNames
+}
+
+func (s s2iComponent) GetEnvVars() []v12.EnvVar {
+	return s.dc.Spec.Template.Spec.Containers[0].Env
+}
+
+func (s s2iComponent) GetLabels() map[string]string {
+	return s.dc.Labels
+}
+
+func (s s2iComponent) GetAnnotations() map[string]string {
+	return s.dc.Annotations
+}
+
+func (s s2iComponent) GetName() string {
+	return s.dc.Labels[componentlabels.ComponentLabel]
+}
+
+func (s s2iComponent) GetType() (string, error) {
+	return getType(s)
+}
+
+func (s s2iComponent) GetSource() (string, string, error) {
+	return getSource(s)
+}
+
+type devfileComponent struct {
+	d v1.Deployment
+}
+
+func (d devfileComponent) GetLinkedSecretNames() (secretNames []string) {
+	for _, container := range d.d.Spec.Template.Spec.Containers {
+		for _, env := range container.EnvFrom {
+			if env.SecretRef != nil {
+				secretNames = append(secretNames, env.SecretRef.Name)
+			}
+		}
+	}
+	return secretNames
+}
+
+func (d devfileComponent) GetEnvVars() []v12.EnvVar {
+	var envs []v12.EnvVar
+	for _, container := range d.d.Spec.Template.Spec.Containers {
+		envs = append(envs, container.Env...)
+	}
+	return envs
+}
+
+func (d devfileComponent) GetLabels() map[string]string {
+	return d.d.Labels
+}
+
+func (d devfileComponent) GetAnnotations() map[string]string {
+	return d.d.Annotations
+}
+
+func (d devfileComponent) GetName() string {
+	return d.d.Name
+}
+
+func (d devfileComponent) GetType() (string, error) {
+	return getType(d)
+}
+
+func (d devfileComponent) GetSource() (string, string, error) {
+	return getSource(d)
+}
+
+type noSourceError struct {
+	msg string
+}
+
+func (n noSourceError) Error() string {
+	return n.msg
+}
+
+func getSource(component provider) (string, string, error) {
+	annotations := component.GetAnnotations()
+	if sourceType, ok := annotations[ComponentSourceTypeAnnotation]; ok {
+		if !validateSourceType(sourceType) {
+			return "", "", fmt.Errorf("unsupported component source type %s", sourceType)
+		}
+		var sourcePath string
+		if sourceType == string(config.GIT) {
+			sourcePath = annotations[componentSourceURLAnnotation]
+		}
+
+		klog.V(4).Infof("Source for component %s is %s (%s)", component.GetName(), sourcePath, sourceType)
+		return sourceType, sourcePath, nil
+	}
+	return "", "", noSourceError{msg: fmt.Sprintf("%s component doesn't provide a source type annotation", component.GetName())}
+}
+
+func getType(component provider) (string, error) {
+	if componentType, ok := component.GetLabels()[componentlabels.ComponentTypeLabel]; ok {
+		return componentType, nil
+	}
+	return "", fmt.Errorf("%s component doesn't provide a type label", component.GetName())
+}
+
+func GetPushedComponents(c *occlient.Client, applicationName string) (map[string]PushedComponent, error) {
```

Comment:
Any reason why we are returning a list instead of a map?

Reply:
Please add some comment for the function

Reply:
Why not return a map?

Reply:
ordering becomes predictable

Reply:
Because this function is only called by `GetComponentNames()` which loops over the map keys to get the names. But we can loop over the list of components and use `<pushed-component>.GetName()` to get the name of each component too. Unless we use the keys to find individual items, a map usage feels weird to me as we are looping over all the items.

Reply:
Sure, but there is no intrinsic order here which returning a list would imply. `GetComponentNames` should indeed return a sorted list, which I will do.